### PR TITLE
[SYCL][ESIMD] Update ESIMD tests and add raw send support.

### PR DESCRIPTION
**Pull request by kychendev**
Update ESIMD tests and add raw send support.

### CODE REVIEW

**Comment by romanovvlad** [comment]
@kbobrovs  @rbegam @DenisBakhvalov  Could you please review?

File: sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp [hunk 1]
```diff
@@ -643,6 +643,198 @@ media_block_store(AccessorTy acc, unsigned x, unsigned y, simd<T, m * n> vals) {
 SYCL_EXTERNAL void slm_init(uint32_t size) {}
 
 #endif
+
+template <typename AccessorTy>
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp:647]: please document this API

File: sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp [hunk 2]
```diff
@@ -643,6 +643,198 @@ media_block_store(AccessorTy acc, unsigned x, unsigned y, simd<T, m * n> vals) {
...
+ESIMD_INLINE ESIMD_NODEBUG uint32_t esimd_get_value(AccessorTy acc) {
+#if defined(__SYCL_DEVICE_ONLY__) && defined(__SYCL_EXPLICIT_SIMD__)
+  return __esimd_get_value(AccessorPrivateProxy::getNativeImageObj(acc));
+#else
+  return __esimd_get_value(acc);
+#endif // __SYCL_DEVICE_ONLY__ && __SYCL_EXPLICIT_SIMD__
+}
+
+/// \brief Raw sends load.
+///
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp:657]: Please provide some comments what is "Raw send".
Does it model "send" instruction from https://01.org/sites/default/files/documentation/intel-gfx-prm-osrc-icllp-vol02a-commandreference-instructions_2.pdf? Maybe it makes sense to add a link then?

File: sycl/include/CL/sycl/INTEL/esimd/esimd_math.hpp [hunk 3]
```diff
@@ -1838,7 +1838,7 @@ template <typename T0, typename T1, int SZ> struct esimd_apply_reduced_min {
 
 template <typename T0, typename T1, int SZ,
           template <typename RT, typename T, int N> class OpType>
-T1 esimd_reduce_single(simd<T1, SZ> v) {
+T0 esimd_reduce_single(simd<T1, SZ> v) {
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/INTEL/esimd/esimd_math.hpp:1841]: This change does not seem to be related to raw_send.

File: sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp [hunk 4]
```diff
@@ -643,6 +643,198 @@ media_block_store(AccessorTy acc, unsigned x, unsigned y, simd<T, m * n> vals) {
...
+/// @msgSrc1 the second source operand of send message.
+///
+/// @exDesc the extended message descriptor.
+///
+/// @msgDesc the message descriptor.
+///
+/// @execSize the execution size, which must be a compile time constant.
+///
+/// @sfid the shared function ID, which must be a compile time constant.
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp:670]: the API would be hard to use w/o some more info on what exDesc, msgDesc and sfid can be.

File: sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp [hunk 5]
```diff
@@ -643,6 +643,198 @@ media_block_store(AccessorTy acc, unsigned x, unsigned y, simd<T, m * n> vals) {
...
+/// constant.
+///
+/// @isEOT 	the flag that indicates whether this is an EOT message, which
+/// must be a compile time constant (optional - default to 0).
+///
+/// @isSendc the flag that indicates whether sendc should be used, which must be
+/// a compile time constant (optional - default to 0).
+///
+/// @mask the predicate to specify enabled channels (optional - default to on).
+///
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp:688]: Please add info what this operation returns

File: sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp [hunk 6]
```diff
@@ -643,6 +643,198 @@ media_block_store(AccessorTy acc, unsigned x, unsigned y, simd<T, m * n> vals) {
...
+  return __esimd_get_value(AccessorPrivateProxy::getNativeImageObj(acc));
+#else
+  return __esimd_get_value(acc);
+#endif // __SYCL_DEVICE_ONLY__ && __SYCL_EXPLICIT_SIMD__
+}
+
+/// \brief Raw sends load.
+///
+/// @msgDst the destination operand of send message.
+///
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp:659]: it is not clear how this parameter is to be used - it is passed by value, yet it is a destination.

**Review by kbobrovs — CHANGES_REQUESTED** [review]

File: sycl/include/CL/sycl/INTEL/esimd/esimd_math.hpp (hunk 3 above)
Comment (kychendev) [reply sycl/include/CL/sycl/INTEL/esimd/esimd_math.hpp:1841]: Right, as you can see from the change list, this patch includes other local changes including ESIMD tests update and a reduction test fix.

File: sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp (hunk 1 above)
Comment (kychendev) [reply sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp:653]: Added comment.

File: sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp (hunk 2 above)
Comment (kychendev) [reply sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp:663]: Yes, this API is used to implement the send messages. Added link to the open source document.

File: sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp (hunk 6 above)
Comment (kychendev) [reply sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp:665]: It's the old value of the destination operand - updated comment to clarify this.

File: sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp (hunk 5 above)
Comment (kychendev) [reply sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp:701]: Added comment.

File: sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp (hunk 4 above)
Comment (kychendev) [reply sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp]: These correspond to the different send message encoding fields and are described in the open source reference document (added link in the comment). This API is intended for expert programmers who will need to read the reference manual in order to program the desired send message function.

**Comment by romanovvlad** [comment]
@kbobrovs @rbegam  Could you please re-review?

File: sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp [hunk 7]
```diff
@@ -643,6 +643,216 @@ media_block_store(AccessorTy acc, unsigned x, unsigned y, simd<T, m * n> vals) {
...
+/// @param isEOT the flag that indicates whether this is an EOT message, which
+/// must be a compile time constant (optional - default to 0).
+///
+/// @param isSendc the flag that indicates whether sendc should be used, which
+/// must be a compile time constant (optional - default to 0).
+///
+/// @param mask the predicate to specify enabled channels (optional - default to
+/// on).
+///
+/// Returns a simd vector of type T1 and size n1.
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp:696]:
```suggestion
/// Returns the vector value read from the memory.
```

Comment (kbobrovs) [inline sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp:696]:
```suggestion
/// Returns the vector value read from memory.
```

File: sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp [hunk 8]
```diff
@@ -643,6 +643,216 @@ media_block_store(AccessorTy acc, unsigned x, unsigned y, simd<T, m * n> vals) {
...
+///
+/// Raw send APIs are used to implement the send messages on Intel� processor
+/// graphics.
+/// https://01.org/sites/default/files/documentation/intel-gfx-prm-osrc-icllp-vol02a-commandreference-instructions_2.pdf
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp:700]:
```suggestion
```

File: sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp [hunk 9]
```diff
@@ -643,6 +643,216 @@ media_block_store(AccessorTy acc, unsigned x, unsigned y, simd<T, m * n> vals) {
...
+///
+template <typename AccessorTy>
+ESIMD_INLINE ESIMD_NODEBUG uint32_t esimd_get_value(AccessorTy acc) {
+#if defined(__SYCL_DEVICE_ONLY__) && defined(__SYCL_EXPLICIT_SIMD__)
+  return __esimd_get_value(AccessorPrivateProxy::getNativeImageObj(acc));
+#else
+  return __esimd_get_value(acc);
+#endif // __SYCL_DEVICE_ONLY__ && __SYCL_EXPLICIT_SIMD__
+}
+
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp:661]:
```suggestion
/// Raw send APIs below are used to implement the send messages on Intel(R) processor
/// graphics, as defined in the documentation at
/// https://01.org/sites/default/files/documentation/intel-gfx-prm-osrc-icllp-vol02a-commandreference-instructions_2.pdf

```

File: sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp [hunk 10]
```diff
@@ -643,6 +643,216 @@ media_block_store(AccessorTy acc, unsigned x, unsigned y, simd<T, m * n> vals) {
...
+/// @param isEOT the flag that indicates whether this is an EOT message, which
+/// must be a compile time constant (optional - default to 0).
+///
+/// @param isSendc the flag that indicates whether sendc should be used, which
+/// must be a compile time constant (optional - default to 0).
+///
+/// @param mask the predicate to specify enabled channels (optional - default to
+/// on).
+///
+/// Returns a simd vector of type T1 and size n1.
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp:752]:
```suggestion
/// Returns the vector value read from memory.
```

**Review by kbobrovs — CHANGES_REQUESTED** [review]

**Review by kbobrovs — DISMISSED** [review]

**Comment by kbobrovs** [comment]
@kychendev, I took the liberty to commit few minor comments fixes to approve and speed up the process. Feel free to roll back if needed.

File: sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp [hunk 11]
```diff
@@ -643,6 +643,216 @@ media_block_store(AccessorTy acc, unsigned x, unsigned y, simd<T, m * n> vals) {
...
+template <typename AccessorTy>
+ESIMD_INLINE ESIMD_NODEBUG uint32_t esimd_get_value(AccessorTy acc) {
+#if defined(__SYCL_DEVICE_ONLY__) && defined(__SYCL_EXPLICIT_SIMD__)
+  return __esimd_get_value(AccessorPrivateProxy::getNativeImageObj(acc));
+#else
+  return __esimd_get_value(acc);
+#endif // __SYCL_DEVICE_ONLY__ && __SYCL_EXPLICIT_SIMD__
+}
+/// Raw send APIs below are used to implement the send messages on Intel(R) processor
+/// graphics, as defined in the documentation at
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp:662]:
```suggestion

+/// Raw send APIs below are used to implement the send messages on Intel(R)
+/// processor graphics, as defined in the documentation at
```

**Comment by kychendev** [comment]
@kbobrovs, LGTM. However I see 4 checks failing now, which were passing before. Could you help double check if it's test env issue, or what needs to be changed?

File: sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp [hunk 12]
```diff
@@ -658,8 +658,9 @@ ESIMD_INLINE ESIMD_NODEBUG uint32_t esimd_get_value(AccessorTy acc) {
   return __esimd_get_value(acc);
 #endif // __SYCL_DEVICE_ONLY__ && __SYCL_EXPLICIT_SIMD__
 }
-/// Raw send APIs below are used to implement the send messages on Intel(R) processor
-/// graphics, as defined in the documentation at
+
++/// Raw send APIs below are used to implement the send messages on Intel(R)
++/// processor graphics, as defined in the documentation at
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp:663]:
```suggestion
/// Raw send APIs below are used to implement the send messages on Intel(R)
/// processor graphics, as defined in the documentation at
```

**Comment by kbobrovs** [comment]
@kychendev, sorry for typo - restarted the checks.

**Review by kbobrovs — APPROVED** [review]

**Comment by kbobrovs** [comment]
@intel/llvm-reviewers-runtime , ping. Please review/approve

**Review by romanovvlad — APPROVED** [review]
